### PR TITLE
Unary ObjectTable for RefCounter

### PR DESCRIPTION
**Pull request by mjp41**
Rather than storing the count, we use an entry per reference. This
means it is very quick to add a reference, we simple put the object
in the next slot of the array.

Releasing a reference, is a little harder, we scan the array looking
for the entry, once we find it we can collapse the list into a
condensed form, by moving the first element into the gap, and nudging
the first up.

The API has changed to only return a bool for whether a reference
exists or not. We don't care how many references there are. This means
we can short cut several searchs as soon as we find something.

### CODE REVIEW

File: src/System.Slices/System/Buffers/ReferenceCounter.cs [hunk 1]
```diff
@@ -41,26 +42,25 @@ public void Release(object obj)
...
+        public bool HasReference(object obj)
         {
             var allTables = s_allTables;
             lock (allTables)
             {
-                uint globalCount = 0;
                 for (int index = 0; index < s_allTablesCount; index++)
                 {
-                    globalCount += allTables[index].GetCount(obj);
+                    if(allTables[index].HasReference(obj)) return true;
```
Comment (davidfowl) [inline src/System.Slices/System/Buffers/ReferenceCounter.cs]: nit: formatting, format the file (space between if and ( )

**Comment by mjp41** [comment]
@KrzysztofCwalina I've modified #1047.  Assuming, we don't have many references to the same object from a single thread, this looks to be about 50% faster.

**Comment by KrzysztofCwalina** [comment]
@mjp41, do you have any benchmark results showing the impact of the change?

**Comment by mjp41** [comment]
@KrzysztofCwalina,  Nothing very comprehensive yet. I just used the Slices.Tests, I made the two thread example you have in the test harness, and increased the runs by 100x. That test, took 2.8 seconds with the original code, and 2 seconds with the new code. So the original code was 40% slower (not 50% like I said in my earlier comment).

I'll try to perform more comprehensive benchmarking tonight.

File: src/System.Slices/System/Buffers/ReferenceCounter.cs [hunk 2]
```diff
@@ -85,129 +85,86 @@ private static ObjectTable AddThreadLocalTable()
...
+                var oldLength = _items.Length;
+                var larger = new object[oldLength << 1];
+                _items.CopyTo(larger, oldLength);
+                // The copy must occur before the table is published.
+                Volatile.Write(ref _items, larger);
+                // _items must be updated before the value of _first is changed
+                // If this is not the case, then we could effectively make the
+                // table empty temporarily.
+                _items[oldLength - 1] = obj;
+                Volatile.Write(ref _first, oldLength - 1);
```
Comment (davidfowl) [inline src/System.Slices/System/Buffers/ReferenceCounter.cs]: Is the Volatile.Write necessary?

**Comment by KrzysztofCwalina** [comment]
@mjp41, I ran the program/benchmark below and got the following results:

Interlocked                30594ms
ReferenceCounter         320ms
ReferenceCounterPR    493ms

I would prefer to use BenchmarkDotNet, but it seems to have problems with the thread static field.

```c#
class Benchmark
{
    static Stopwatch watch = new Stopwatch();
    public static void Run()
    {
        int itterations = 10000000;
        int threads = 20;

        // warmup
        StopwatchBenchInterlocked(true, 1, 2);
        StopwatchBenchInterlocked(false, itterations, threads);

        // warmup
        StopwatchBenchReferenceCounter(true, 1, 2);
        StopwatchBenchReferenceCounter(false, itterations, threads);

        // warmup
        StopwatchBenchReferenceCounter2(true, 1, 2);
        StopwatchBenchReferenceCounter2(false, itterations, threads);

        Console.ReadLine();
    }

    static void StopwatchBenchReferenceCounter(bool warmup, int itterations, int threads)
    {
        watch.Restart();
        var counter = new ReferenceCounter();
        var obj = new object();
        var tasks = new List<Task>(threads);
        for (int i = 0; i < threads; i++)
        {
            tasks.Add(Task.Run(() => {
                for (int itteration = 0; itteration < itterations; itteration++)
                {
                    counter.AddReference(obj);
                    counter.Release(obj);
                }
            }));
        }

        Task.WaitAll(tasks.ToArray());
        if (!warmup) Console.WriteLine("ReferenceCounter {0}ms", watch.ElapsedMilliseconds);
        if (counter.GetGlobalCount(obj) != 0)
        {
            throw new Exception();
        }
    }
 ...
```

**Comment by KrzysztofCwalina** [comment]
If, I AddRef/Release more than once (20 times, increments == 20), the results are a wash:

ReferenceCounterPR 2998ms
ReferenceCounter 2982ms

```c#
static void StopwatchBenchReferenceCounter(int itterations, int increments, int threads, bool warmup)
    {
        watch.Restart();
        var counter = new ReferenceCounter();
        var obj = new object();
        var tasks = new List<Task>(threads);
        for (int i = 0; i < threads; i++)
        {
            tasks.Add(Task.Run(() => {
                for (int itteration = 0; itteration < itterations; itteration++)
                {
                    for (int increment=0; increment < increments; increment++) counter.AddReference(obj);
                    for (int increment = 0; increment < increments; increment++) counter.Release(obj);
                }
            }));
        }

        Task.WaitAll(tasks.ToArray());
        if (!warmup) Console.WriteLine("ReferenceCounter {0}ms", watch.ElapsedMilliseconds);
        if (counter.GetGlobalCount(obj) != 0)
        {
            throw new Exception();
        }
    }
```

**Comment by KrzysztofCwalina** [comment]
I also did one object per thread and one object per all threads runs:

**** One Object ****
Interlocked      3127ms
ReferenceCounterPR 29ms
ReferenceCounter 29ms

**** Object Per Thread ****
Interlocked      1469ms
ReferenceCounterPR 29ms
ReferenceCounter 28ms

And the test "template":
```c#
    static void StopwatchBenchReferenceCounter(int itterations, int increments, int threads, bool objectPerThread, bool warmup)
    {
        watch.Restart();
        var counter = new ReferenceCounter();
        var obj = new object();
        var tasks = new List<Task>(threads);
        for (int i = 0; i < threads; i++)
        {
            tasks.Add(Task.Run(() => {
                var threadObject = objectPerThread ? new object() : obj;
                for (int itteration = 0; itteration < itterations; itteration++)
                {
                    for (int increment=0; increment < increments; increment++) counter.AddReference(threadObject);
                    for (int increment = 0; increment < increments; increment++) counter.Release(threadObject);
                }
            }));
        }

        Task.WaitAll(tasks.ToArray());
        if (!warmup) Console.WriteLine("ReferenceCounter {0}ms", watch.ElapsedMilliseconds);
        if (counter.GetGlobalCount(obj) != 0)
        {
            throw new Exception();
        }
    }
```

**Comment by KrzysztofCwalina** [comment]
cc: @jkotas, @vancem

**Comment by mjp41** [comment]
@krzysztofCwalina, thanks for the benchmark.  I realized my earlier runs were with Debug build :(.  This might explain the difference in performance,

Removing the Volatile.Writes improves performance, but I am concerned the compiler could perform to many reorderings if I do that. 

Do you think the single object scenario is the common one? What else would be required for ASP?  I was planning to benchmark multiple objects with scoped lifetimes.

**Comment by KrzysztofCwalina** [comment]
> Do you think the single object scenario is the common one? What else would be required for ASP? I was planning to benchmark multiple objects with scoped lifetimes.

Yeah, let me try to improve the benchmark to simulate multiple objects per thread.

**Comment by KrzysztofCwalina** [comment]
I added the ability to run it with multiple objects being refcounted. Here are the results:

unning on ProcessorCount threads
**** One Object Per Process ****
ReferenceCounterPR 323ms
ReferenceCounter 345ms

**** One Object Per Thread ****
ReferenceCounterPR 332ms
ReferenceCounter 343ms

**** 4 Objects Per Thread ****
ReferenceCounterPR 1052ms
ReferenceCounter 845ms

**** 10 Objects Per Thread ****
ReferenceCounterPR 2603ms
ReferenceCounter 2818ms

I think your counter is winning consistently for 8+ objects. 

Here is the new code:

```c#
 static void StopwatchBenchReferenceCounter(int itterations, int increments, int threads, int objectsPerThread, bool warmup)
    {
        watch.Restart();
        var counter = new ReferenceCounter();
        var singleton = new object();
        var tasks = new List<Task>(threads);
        for (int i = 0; i < threads; i++)
        {
            tasks.Add(Task.Run(() => {
                object[] objects = CreateObjectArray(objectsPerThread, singleton);

                for (int itteration = 0; itteration < itterations; itteration++)
                {
                    for (int increment = 0; increment < increments; increment++)
                    {
                        for (int objectIndex = 0; objectIndex < objects.Length; objectIndex++)
                        {
                            counter.AddReference(objects[objectIndex]);
                        }
                    }
                    for (int increment = 0; increment < increments; increment++)
                    {
                        for (int objectIndex = 0; objectIndex < objects.Length; objectIndex++)
                        {
                            counter.Release(objects[objectIndex]);
                        }
                    }
                }
            }));
        }

        Task.WaitAll(tasks.ToArray());
        if (!warmup) Console.WriteLine("ReferenceCounter {0}ms", watch.ElapsedMilliseconds);
    }
```

**Comment by mattwarren** [comment]
> I would prefer to use BenchmarkDotNet, but it seems to have problems with the thread static field.

@KrzysztofCwalina Is a gist available of the BenchmarkDotNet code (that didn't work)? I'll can take a look and see what needs to be fixed in BenchmarkDotNet to make it work

File: src/System.Slices/System/Buffers/ReferenceCounter.cs (hunk 2 above)
Comment (mjp41) [inline src/System.Slices/System/Buffers/ReferenceCounter.cs]: I believe it is, if the write at 119 was to move above 115, then the effect would be make the array appear to contain just one element, until the update to _items is propogated.  Now, I am not sure if it is something that any implementations do, but I thought it was allowed by the ECMA spec.

**Comment by mjp41** [comment]
@KrzysztofCwalina, Re:Interlocked.  if most of the use cases are single threaded access to a memory, then it might be more sensible to use an Interlocked approach.

It looks like the distributed reference counting would be more predicable in performance, which may be worth the additional cost in the single threaded scenario.

**Comment by mjp41** [comment]
I have run a version of the benchmark @KrzysztofCwalina posted above. 

```
RefCounter	RefCounterPR	Diff		incs	threads	objects	stack
26.51538846	21.95148873	4.563899728	1	1	1	False
50.48577471	44.98663581	5.499138898	2	1	1	False
273.4114092	246.2657649	27.14564427	10	1	1	False
51.60448573	46.41327265	5.191213083	1	4	1	False
97.97350931	87.30230329	10.67120603	2	4	1	False
563.5884138	470.6203014	92.96811243	10	4	1	False
126.39318	122.0593254	4.333854646	1	10	1	False
241.5870796	215.5810658	26.00601381	2	10	1	False
1391.242897	1089.52205	301.7208464	10	10	1	False
47.36275391	35.17149372	12.19126018	1	1	2	False
85.16737003	71.83364905	13.33372098	2	1	2	False
416.6588034	361.5787794	55.08002398	10	1	2	False
76.46031448	66.24595719	10.21435729	1	4	2	False
135.9384363	131.8118005	4.126635862	2	4	2	False
696.7416486	654.5310468	42.21060177	10	4	2	False
211.3519516	168.3692613	42.98269025	1	10	2	False
396.7915127	335.3701213	61.42139134	2	10	2	False
1720.556606	1689.982105	30.57450126	10	10	2	False
300.659857	119.7201875	180.9396694	1	1	10	False
593.3997575	256.3407632	337.0589944	2	1	10	False
2714.165427	1259.078412	1455.087015	10	1	10	False
484.9252464	250.8020469	234.1231994	1	4	10	False
979.7013509	453.6595688	526.0417821	2	4	10	False
4370.204714	2185.218272	2184.986442	10	4	10	False
1318.177902	588.3872682	729.7906339	1	10	10	False
2526.928925	1175.490482	1351.438442	2	10	10	False
11138.29714	5506.949202	5631.347937	10	10	10	False
27.2467548	22.5438799	4.702874899	1	1	1	True
55.18775111	43.36224424	11.82550687	2	1	1	True
284.1731729	237.5528383	46.6203346	10	1	1	True
56.95177207	43.47652363	13.47524844	1	4	1	True
101.5980198	89.14550185	12.45251793	2	4	1	True
478.6905373	450.1867284	28.50380894	10	4	1	True
126.088886	113.1007196	12.98816639	1	10	1	True
289.6904143	212.3413317	77.34908261	2	10	1	True
1244.81146	1142.767871	102.0435892	10	10	1	True
47.23749343	31.0519026	16.18559084	1	1	2	True
83.60352696	60.58059219	23.02293477	2	1	2	True
417.534439	327.1650249	90.36941416	10	1	2	True
76.4498778	65.74343057	10.70644724	1	4	2	True
139.1053955	120.8528366	18.25255884	2	4	2	True
687.8645131	602.7425344	85.12197867	10	4	2	True
196.0739761	151.5933215	44.48065459	1	10	2	True
372.4552653	305.9194184	66.53584688	2	10	2	True
1755.234261	1474.5328	280.7014611	10	10	2	True
291.5802506	88.27709183	203.3031588	1	1	10	True
563.2808077	178.3014089	384.9793988	2	1	10	True
2644.090046	918.0411977	1726.048849	10	1	10	True
464.0785436	160.7682217	303.3103219	1	4	10	True
894.0096645	311.6868922	582.3227723	2	4	10	True
4600.134236	1635.589957	2964.544279	10	4	10	True
1236.103578	400.2238708	835.8797075	1	10	10	True
2254.885542	803.4742124	1451.41133	2	10	10	True
10563.52518	4102.270033	6461.255148	10	10	10	True
```
On my machine it shows wins across the board. Stack indicates if the release is FIFO or LIFO.

**Comment by KrzysztofCwalina** [comment]
Sounds good. Let's merge it.
Thanks!

**Pull request merged**